Realm.Shared/linq/RealmResultsVisitor.cs

```diff
@@ -179,44 +216,60 @@ internal override Expression VisitMethodCall(MethodCallExpression m)
                             firstRowPtr = rh.GetRow(0);
                         }
                     }
-                    if (firstRowPtr == IntPtr.Zero)
-                        throw new InvalidOperationException("Sequence contains no matching element");
-                    return Expression.Constant(_realm.MakeObjectForRow(_metadata, firstRowPtr));
+                    if (firstRowPtr != IntPtr.Zero)
+                        return Expression.Constant(_realm.MakeObjectForRow(_metadata, firstRowPtr));
+                    if (m.Method.Name == "First")
+                            throw new InvalidOperationException("Sequence contains no matching element");
+                    Debug.Assert (m.Method.Name == "FirstorDefault");
+                    return MakeDefaultWithTransactionIfNeeded(_metadata);
```

Comment:
Not sure if I understand it correctly, but it seems here, we'll create a new object instead of returning null or am I missing something?